doc/design/context.md

```diff
@@ -0,0 +1,171 @@
+## Context Design
+
+`Net` is the container and controller of a set of `Operator`. Each `Operator` in `Net` has a method called `Run` to make computation. The `Run` method of `Operator` is defined as follows:
+
+```c++
+Error Operator::Run(OpContext* context);
+```
+
+And the `OpContext` is defined as follows:
+
+```c++
+Struct OpContext {
+  Scope* scope;
+  DeviceContext* device_context;
+};
+```
+
+The `Run` method will take `Variable` (containing `Tensor`) from `Scope` to make computation on certain `DeviceContext`. `DeviceContext` provides necessary runtime resources for computation, including CudaStream, cublasHandle and so on.
```

Comment:
Is it something like `Operator::Run` can call `Operator::Input(i)` to get the i-th input, which should have been created by the feeding operator or some other one before the invocation of `Operator::Run`?

If so, it seems that `Operator::Run` has to run on the same device as where the input resides; otherwise, there would be unnecessary data copying from where the input has been to where `Run` is on.

Given that we don't want such kind of inefficient copying, we should make sure that all operators of a network run on the same device, and `Operator::Run` should take a constant reference to the context, like

```cpp
void Operator::Run(const Context& ctx);
```

and Context should refer to a single specific device (like GPU) other than a set of GPUs?

I think we can do so actually, just `class Net` a simple lightweight class without sub-classes.  If we want to make use of multiple GPUs, we can create `class DataParallelEngine` which runs multiple clones of a net, each on a GPU and by a thread, and aggregates gradient tensors of model parameters.  A simpler engine could be `class SingleThreadEngine` which runs a net on a specific device using one thread.